pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			tpcc.run(ctx, u.c, d)
+			return nil
+		})
+	}
+}
+
+// waitForVerifier waits for the underlying CDC verifier to reach the
+// desired number of resolved timestamps.
+func (cmvt *cdcMixedVersionTester) waitForVerifier() versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("waiting for verifier")
+		<-cmvt.verifierDone
+	}
+}
+
+// setupVerifier creates a CDC validator to validate that a changefeed
+// created on the `tpcc.warehouse` table is able to re-create the
+// table somewhere else. It also verifies CDC's ordering
+// guarantees. This step will not block, but will start the verifier
+// in a separate Go routine. Use `waitForVerifier` to wait for
+// the verifier to finish.
+func (cmvt *cdcMixedVersionTester) setupVerifier(node int) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status(fmt.Sprintf("setting up changefeed verifier for table %s", target))
+
+		// we could just return the error here and let `Wait` return the
+		// error. However, calling t.Fatal directly lets us stop the test
+		// earlier
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			consumer, err := cmvt.kafka.consumer(ctx, "warehouse")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer consumer.Close()
+
+			db := u.conn(ctx, t, node)
+			if _, err := db.Exec(
+				"CREATE TABLE fprint (" +
+					"w_id INT NOT NULL PRIMARY KEY, " +
+					"w_name VARCHAR(10) NOT NULL, " +
+					"w_street_1 VARCHAR(20) NOT NULL, " +
+					"w_street_2 VARCHAR(20) NOT NULL, " +
+					"w_city VARCHAR(20) NOT NULL, " +
+					"w_state VARCHAR(2) NOT NULL, " +
+					"w_zip VARCHAR(9) NOT NULL, " +
+					"w_tax DECIMAL(4, 4) NOT NULL, " +
+					"w_ytd DECIMAL(12, 2) NOT NULL)",
+			); err != nil {
+				t.Fatal(err)
+			}
+
+			fprintV, err := cdctest.NewFingerprintValidator(db, `tpcc.warehouse`, `fprint`, consumer.partitions, 0)
+			if err != nil {
+				t.Fatal(err)
+			}
+			validators := cdctest.Validators{
+				cdctest.NewOrderValidator("tpcc.warehouse"),
+				fprintV,
+			}
+			cmvt.validator = cdctest.MakeCountValidator(validators)
+
+			for {
+				m := consumer.Next(ctx)
+				if m == nil {
+					t.Fatal(fmt.Errorf("unexpected end of changefeed"))
+					return nil
+				}
+
+				updated, resolved, err := cdctest.ParseJSONValueTimestamps(m.Value)
+				if err != nil {
+					t.Fatal(err)
+					return nil
+				}
+
+				partitionStr := strconv.Itoa(int(m.Partition))
+				// errors in the calls to the validator below could lead to
+				// transient failures if the cluster is upgrading, so we retry
+				// for 1 minute.
+				if len(m.Key) > 0 {
+					if err := retry.ForDuration(1*time.Minute, func() error {
+						return cmvt.validator.NoteRow(partitionStr, string(m.Key), string(m.Value), updated)
+					}); err != nil {
+						t.Fatal(err)
+						return nil
+					}
+				} else {
+					if err := retry.ForDuration(1*time.Minute, func() error {
```

Comment:
Okay, I haven't proved it locally yet but I think this right here is the source of the flakes. The FingerprintValidator's NoteResolved method is neither atomic nor idempotent, so if it hits an error when trying to write to a restarting database, it will end up in an inconsistent state. It has an in-memory buffer of consumed events to be applied (populated by NoteRow), and iterates over it by doing 
```
row := v.buffer[0]
v.buffer = v.buffer[1:]
if err := v.applyRowUpdate(row); err != nil {
			return err
		}
```

So if the attempt to write a row to the scratch table in applyRowUpdate fails, we'll never actually retry the write with that row.

A simple solution might be to move the retry logic down into the actual query execution? Or we could make the validator methods more ACID.

Reply:
Just changing FingerprintValidator's NoteResolved to only modify the validator state when it's about to return a nil error might do it, but I haven't audited all the validator methods, there might be one or two similar bugs there.

Reply:
That's indeed a very possible reason for this, let me run some tests. 👍

Reply:
This does seem to fix the issue of the fingerprint validator failing in this test, good catch @HonoreDB!

Reply:
First commit in this PR shifts the mutation on the validator to _after_ the `applyRowUpdate` call (which, afaict, is the only call in the validator that writes to a table). I haven't seen fingerprint mismatch errors after the change.